compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala

```diff
@@ -142,7 +142,8 @@ trait TypeAssigner {
     val mbr = qualType.findMember(name, pre)
     if (reallyExists(mbr))
       qualType.select(name, mbr)
-    else if (qualType.derivesFrom(defn.DynamicClass) && name.isTermName && !Dynamic.isDynamicMethod(name))
+    else if (qualType.derivesFrom(defn.DynamicClass) && name.isTermName && !Dynamic.isDynamicMethod(name) &&
+        !(name == nme.apply && Dynamic.isDynamicMethod(qual1.symbol.name)))
```

Comment:
The following code type checks before, but fails to type-check with the change:

```Scala
import scala.language.dynamics

class Foo extends Dynamic {
  def applyDynamic(name: String)(): Bar = ???
}

class Bar extends Dynamic {
  def applyDynamic(name: String)(x: Int) = ???
}

val foo = new Foo
def baz = foo.blah().apply(42)
```

Reply:
Good catch. I need to check that the `apply` was not written in the source.

Reply:
Fixed